### PR TITLE
fix `_checked_mul_dims` in the presence of 0s and overflow.

### LINKED ISSUE #54244: whether `Core.checked_dims` decides to throw depends on dimension order (for empty arrays)
```julia-repl
julia> b = typemax(Int) - 1
9223372036854775806

julia> Array{Int}(undef, b, b, 0)
ERROR: ArgumentError: invalid Array dimensions
Stacktrace:
 [1] checked_dims
   @ ./boot.jl:585 [inlined]
 [2] Array
   @ ./boot.jl:598 [inlined]
 [3] (Array{Int64})(::UndefInitializer, m::Int64, n::Int64, o::Int64)
   @ Core ./boot.jl:611
 [4] top-level scope
   @ REPL[2]:1

julia> Array{Int}(undef, 0, b, b)
0×9223372036854775806×9223372036854775806 Array{Int64, 3}

julia> versioninfo()
Julia Version 1.12.0-DEV.387
Commit b5bfd83a3d0 (2024-04-22 13:12 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × AMD Ryzen 3 5300U with Radeon Graphics
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, znver2)
Threads: 1 default, 0 interactive, 1 GC (on 8 virtual cores)
```

Two alternative solutions:

1. When there's a zero among the dimension sizes, only check that the nonzero sizes are nonnegative and less than `typemax(Int)`. Don't check for overflow during multiplication. This would make both examples above return.

2. Remove the zero sizes before checking for overflow. This would make both examples above throw.

### PR DESCRIPTION
**Pull request by oscardssmith**
fixes https://github.com/JuliaLang/julia/issues/54244.

### CODE REVIEW

**Review by vtjnash — CHANGES_REQUESTED** [review]
This appears to compute overflow wrong in the case where some of the values are negative

**Review by vtjnash — COMMENTED** [review]
Aren't you masking the result unconditionally?

**Comment by oscardssmith** [comment]
Turns out there was already a bug here!
```
julia> zeros(-4, 2, 0)
-4×2×0 Array{Float64, 3}
```

**Comment by nsajko** [comment]
I think this code allows `typemax(Int)` as the size along a dimension, if the size along some other dimension is zero, and no size is negative. Is that intentional? FTR this fails with `ArgumentError` on all recent Julia releases: `Array{Int}(undef, 0, typemax(Int))`, from v1.8 to the v1.11 beta1, but succeeds on v1.7.

**Comment by oscardssmith** [comment]
It does not allow `typemax(Int)` (because of the `ule_int` check). If we wanted to allow them, we would need to use `ult_int`, but I don't think we do.

File: test/core.jl [hunk 1]
```diff
@@ -7213,6 +7213,20 @@ end
...
+Array{Int}(undef, 0, bignum, bignum)
+Array{Int}(undef, bignum, bignum, 0)
+Array{Int}(undef, bignum, bignum, 0, bignum, bignum)
+# but also test that it does throw if the axes multiply to a multiple of typemax(UInt)
+@test_throws ArgumentError Array{Int, 2}(undef, bignum, bignum)
+@test_throws ArgumentError Array{Int, 2}(undef, 1, bignum, bignum)
+# also test that we always throw erros for negative dims even if other dims are 0 or the product is positive
+@test_throws ArgumentError Array{Int, 2}(undef, 0, -4, -4)
+@test_throws ArgumentError Array{Int, 2}(undef, -4, 1, 0)
+@test_throws ArgumentError Array{Int, 2}(undef, -4, -4, 1)
```
Comment (giordano) [inline test/core.jl:7228]: Matrices have 2 dimensions, not 3, maybe you wanted to use `Array{Int, 3}`?

Comment (oscardssmith) [reply test/core.jl]: oops. That's why you don't blindly copy and paste.

**Review by vtjnash — APPROVED** [review]
Yes, I specifically wanted maxdim+1 to be representable, since otherwise other off-by-one math representations become much harder, and that makes boundschecking slower.

**Pull request merged**